### PR TITLE
Add state check for blocks when using trace_module

### PR DESCRIPTION
**Pull request by Demuirgos**
Fixes Closes Resolves #

DO NOT MERGE UNTILL : https://github.com/NethermindEth/nethermind/pull/6237 is merged

_Please choose one of the keywords above to refer to the issue this PR solves followed by the issue number (e.g. Fixes #000). If no issue number, remove the line. Also, remove everything marked optional that is not applicable. Remove this note after reading._

## Changes

- Adds state checks of availability for blocks in tracing rpc endpoints

#### What types of changes does your code introduce?

- [ ] Bugfix (a non-breaking change that fixes an issue)
- [ ] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [ ] Optimization
- [x] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [ ] Other: _Description_

#### Requires testing

- [x] Yes
- [ ] No

#### If yes, did you write tests?

- [ ] Yes
- [x] No

#### Notes on testing

_Optional. Remove if not applicable._

#### Requires documentation update

- [ ] Yes
- [x] No

_If yes, link the PR to the docs update or the issue with the details labeled `docs`. Remove if not applicable._

#### Requires explanation in Release Notes

- [ ] Yes
- [x] No

_If yes, fill in the details here. Remove if not applicable._

## Remarks

_Optional. Remove if not applicable._

### CODE REVIEW

File: src/Nethermind/Nethermind.JsonRpc/Modules/Trace/TraceRpcModule.cs [hunk 1]
```diff
@@ -40,18 +43,26 @@ public class TraceRpcModule : ITraceRpcModule
         private readonly ILogger _logger;
         private readonly ISpecProvider _specProvider;
         private readonly TimeSpan _cancellationTokenTimeout;
+        private readonly IStateReader _stateReader;
 
-        public TraceRpcModule(IReceiptFinder? receiptFinder, ITracer? tracer, IBlockFinder? blockFinder, IJsonRpcConfig? jsonRpcConfig, ISpecProvider? specProvider, ILogManager? logManager)
+        public TraceRpcModule(IStateReader stateReader, IReceiptFinder? receiptFinder, ITracer? tracer, IBlockFinder? blockFinder, IJsonRpcConfig? jsonRpcConfig, ISpecProvider? specProvider, ILogManager? logManager)
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.JsonRpc/Modules/Trace/TraceRpcModule.cs:48]: `tracer` already has the ability to accept visitors, no need to pass `StateReader` here.

File: src/Nethermind/Nethermind.JsonRpc/Modules/Trace/TraceRpcModule.cs [hunk 2]
```diff
@@ -40,18 +43,26 @@ public class TraceRpcModule : ITraceRpcModule
...
             _tracer = tracer ?? throw new ArgumentNullException(nameof(tracer));
             _blockFinder = blockFinder ?? throw new ArgumentNullException(nameof(blockFinder));
             _jsonRpcConfig = jsonRpcConfig ?? throw new ArgumentNullException(nameof(jsonRpcConfig));
             _specProvider = specProvider ?? throw new ArgumentNullException(nameof(specProvider));
             _logManager = logManager ?? throw new ArgumentNullException(nameof(logManager));
+            _stateReader = stateReader ?? throw new AbandonedMutexException(nameof(stateReader));
             _logger = logManager.GetClassLogger();
             _cancellationTokenTimeout = TimeSpan.FromMilliseconds(_jsonRpcConfig.Timeout);
         }
+        private static bool HasStateForBlock(IStateReader stateReader, BlockHeader header)
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.JsonRpc/Modules/Trace/TraceRpcModule.cs:60]: there is already `HasStateForBlock` extension method, not on tracer though, maybe we can have some interface for traversal and make it on this interface?

**Review by MarekM25 — COMMENTED** [review]
@Demuirgos You have never added PR description, please fix it.
Screenshot from 2023-11-02 12-37-40

**Review by MarekM25 — APPROVED** [review]

**Comment by LukaszRozmej** [comment]
please coordinate with #6237

**Review by LukaszRozmej — APPROVED** [review]

**Review by MarekM25 — APPROVED** [review]

**Pull request merged**